core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryActivity.kt

```diff
@@ -0,0 +1,157 @@
+  private val compositeDisposable = CompositeDisposable()
+
+  private val actionModeCallback: Callback =
+    object : Callback {
+      override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
+        mode.menuInflater.inflate(R.menu.menu_context_delete, menu)
+        return true
+      }
+
+      override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean = false
+
+      override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
+        if (item.itemId == R.id.menu_context_delete) {
+          historyViewModel.actions.offer(UserClickedDeleteSelectedHistoryItems)
+          return true
+        }
+        historyViewModel.actions.offer(Action.ExitActionModeMenu)
+        return false
+      }
+
+      override fun onDestroyActionMode(mode: ActionMode) {
+        historyViewModel.actions.offer(Action.ExitActionModeMenu)
+        actionMode = null
+      }
+    }
+
+  private val historyAdapter: HistoryAdapter by lazy {
+    HistoryAdapter(HistoryItemDelegate(this), HistoryDateDelegate())
+  }
+
+  override fun injection(coreComponent: CoreComponent) {
+    activityComponent.inject(this)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_history)
+    setSupportActionBar(toolbar)
+
+    supportActionBar?.setDisplayHomeAsUpEnabled(true)
+    supportActionBar?.setTitle(string.history)
+
+    recycler_view.adapter = historyAdapter
+    recycler_view.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
+
+    history_switch.setOnCheckedChangeListener { _, isChecked ->
+      historyViewModel.actions.offer(UserClickedShowAllToggle(isChecked))
+    }
+    history_switch.isChecked = sharedPreferenceUtil.showHistoryAllBooks
+
+    compositeDisposable.add(historyViewModel.effects.subscribe { it.invokeWith(this) })
+  }
+
+  override fun onDestroy() {
+    compositeDisposable.clear()
+    super.onDestroy()
+  }
+
+  override fun onCreateOptionsMenu(menu: Menu): Boolean {
+    menuInflater.inflate(R.menu.menu_history, menu)
+    val search = menu.findItem(R.id.menu_history_search).actionView as SearchView
+    search.queryHint = getString(R.string.search_history)
+    search.setOnQueryTextListener(SimpleTextListener {
+      historyViewModel.actions.offer(Filter(it))
+    })
+    historyViewModel.state.observe(this, Observer(::render))
+    return true
+  }
+
+  override fun onOptionsItemSelected(item: MenuItem): Boolean {
+    if (item.itemId == android.R.id.home) {
+      historyViewModel.actions.offer(ExitHistory)
+    }
+    if (item.itemId == R.id.menu_history_clear) {
+      historyViewModel.actions.offer(UserClickedDeleteButton)
+    }
+    return super.onOptionsItemSelected(item)
+  }
+
+  private fun render(state: State) {
+    historyAdapter.items = state.getHistoryListItems()
+    if (!state.isInSelectionState) {
+      actionMode?.finish()
+      history_switch.isEnabled = true
```

Comment:
we could take this out of the conditional, looks a little bit more descriptive to just say
```
history_switch.isEnabled = !state.isInSelectionState
```

also a note on the conditional, favour a positive condition, generally reads cleaner, means less logic flipping in the head and `!` by itself is a sneaky little bugger, sometimes the eyes skim over it which i guess is why kotlin Booleans have `.not()`